### PR TITLE
[FancyLogger] Show link to project outputs

### PR DESCRIPTION
**Pull request by edvilme**
Fixes #

### Context
One of the key features of the old console logger is showing the path to the output file for each project using the format
``` Project -> path/to/output```.
To save space, FancyLogger only shows the filename of the output in a clickable link that redirects to the output file following the same format. 
image

### Changes Made
* Updated `ANSIBuilder.Formatting.Hyperlink` to support [clickable hyperlinks](https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda)
* Updated Regex for matching ANSI codes to also match hyperlinks
* TFMs now show in a white background

### CODE REVIEW

**Comment by rainersigwald** [comment]
This is really slick, but I think it hurts a critical use case: getting a path to an output so I can run it.

I often do

1. `dotnet build`
2. double-click output path message to get full path to a specific output
3. right click (copy)
4. right click (paste)
5. enter (run that thing)

I find this easier than `dotnet run` because you can build a whole solution and because you can iterate on adding args to the thing more quickly and easily.

Fortunately I think there's an easy fix: don't truncate the path, but still hyperlink it. Then I can copy/paste but also see it in Explorer.

File: src/Build/Logging/FancyLogger/FancyLoggerProjectNode.cs
Comment (rainersigwald) [inline src/Build/Logging/FancyLogger/FancyLoggerProjectNode.cs:130]: This definitely accomplishes the goal! However, it's pretty inefficent on .NET Framework 4.8:

1. This method is https://referencesource.microsoft.com/#System/regex/system/text/regularexpressions/Regex.cs,798 which calls
2. https://referencesource.microsoft.com/#System/regex/system/text/regularexpressions/Regex.cs,9aef9195b829acff
3. Which creates a new `Regex` object for each call
4. That object's `Match` method is called, creating a `Match` object
5. The `match.Value` allocates a string to hold the latter half of the message.
6. The `Regex` and `Match` objects become garbage after this AddMessage call

This isn't quite as bad as it might be because `new Regex`  has a cache that will not require it to parse + compile the expression again, but it's still allocatey.

There are a couple of options to fix this. The easiest one is to move the regex to a `static` (so it's never garbage but there's only ever one), but that still allocates a `Match` object.

Since this regex is pretty simple I prefer a more manual but minimal-allocation approach:

```suggestion

            const string FinalOutputMarker = " -> ";
            var i = args.Message!.IndexOf(FinalOutputMarker);
            if (i > 0)
            {
                ProjectOutputExecutable = args.Message!.Substring(i + FinalOutputMarker.Length);
            }
```

The only allocation there is the substring, which could be avoided by preserving a span/memory into the original string, but this will allow the GC to collect the whole `BuildMessageEventArgs`, so I think it's better.

File: src/Build/Logging/FancyLogger/FancyLoggerProjectNode.cs
Comment (rainersigwald) [inline src/Build/Logging/FancyLogger/FancyLoggerProjectNode.cs:30]: IMO this should be

```suggestion
        public string? ProjectOutputExecutable;
```

and assign/check for `null` instead of `Length > 0` elsewhere.

File: src/Build/Logging/FancyLogger/FancyLoggerProjectNode.cs
Comment (edvilme) [reply src/Build/Logging/FancyLogger/FancyLoggerProjectNode.cs:130]: Can there be a case in which a high priority message in this format but **without** the project name at the beginning exist?

**Comment by edvilme** [comment]
> This is really slick, but I think it hurts a critical use case: getting a path to an output so I can run it.
> 
> 
> 
> I often do
> 
> 
> 
> 1. `dotnet build`
> 
> 2. double-click output path message to get full path to a specific output
> 
> 3. right click (copy)
> 
> 4. right click (paste)
> 
> 5. enter (run that thing)
> 
> 
> 
> I find this easier than `dotnet run` because you can build a whole solution and because you can iterate on adding args to the thing more quickly and easily.
> 
> 
> 
> Fortunately I think there's an easy fix: don't truncate the path, but still hyperlink it. Then I can copy/paste but also see it in Explorer.

Alright. Do you think we should clip them if they're too long?

**Comment by rainersigwald** [comment]
> Do you think we should clip them if they're too long?

I'm torn on this but lean toward "no, output the whole path". Would happily receive feedback!

File: src/Build/Logging/FancyLogger/FancyLoggerProjectNode.cs
Comment (rainersigwald) [reply src/Build/Logging/FancyLogger/FancyLoggerProjectNode.cs:130]: Yes, at minimum one could have a target that says `<Message Importance="High" Text="X -> Y" />`, which is basically all the message we're looking for here does:

https://github.com/dotnet/msbuild/blob/5a635cf1218a5d7836114de1e392e7613d569675/src/Tasks/Microsoft.Common.CurrentVersion.targets#L4715

You could either check the initial substring or check whether the parent Target is `CopyFilesToOutputDirectory`.

Comment (edvilme) [reply src/Build/Logging/FancyLogger/FancyLoggerProjectNode.cs:130]: Alright, so I should only check for messages whose parent target is `CopyFilesToOutputDirectory` using `IndexOf` instead of Regex, right?

File: src/Build/Logging/FancyLogger/FancyLoggerBuffer.cs
Comment (rainersigwald) [inline src/Build/Logging/FancyLogger/FancyLoggerBuffer.cs:109]: What does this do, when done in the alt buffer just before switching back?

File: src/Build/Logging/FancyLogger/ANSIBuilder.cs
Comment (edvilme) [reply src/Build/Logging/FancyLogger/ANSIBuilder.cs:48]: I agree. It was using spans previously, but couldn't find a way to get Regex working with them, so I would end up using the `ToString()` method which IMO defeats the purpose. 
I will find a way later to make it more memory efficient and rewrite it as a loop

Comment (rainersigwald) [reply src/Build/Logging/FancyLogger/ANSIBuilder.cs:48]: I think you'll be interested in [the `Regex.Match(string, int)` overload](https://learn.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.regex.match?view=net-7.0#system-text-regularexpressions-regex-match(system-string-system-int32)) which allows you to match a "substring" that's actually a string + a start point.

Comment (edvilme) [reply src/Build/Logging/FancyLogger/ANSIBuilder.cs]: Ohh, that will be very useful! Will definitely check it out

File: src/Build/Logging/FancyLogger/FancyLogger.cs
Comment (edvilme) [reply src/Build/Logging/FancyLogger/FancyLogger.cs:185]: Oh in this case it is because we are not using the alternate `FancyLoggerBuffer` but the normal `Console.WriteLine`

File: src/Build/Logging/FancyLogger/FancyLoggerBuffer.cs
Comment (edvilme) [reply src/Build/Logging/FancyLogger/FancyLoggerBuffer.cs:109]: This moves the cursor to the beginning of the screen (0,0) and deletes everything until the end. THat means, deletes everything in the alternate buffer before switching back. This is due to a bug when switching from and to the alt buffer.

Comment (rainersigwald) [reply src/Build/Logging/FancyLogger/FancyLoggerBuffer.cs:109]: Describe this + the bug it works around in a comment please :)

File: src/Build/Logging/FancyLogger/ANSIBuilder.cs
Comment (edvilme) [reply src/Build/Logging/FancyLogger/ANSIBuilder.cs]: Perhaps this change would be better suited for #8331, but what do you think?

Comment (rainersigwald) [reply src/Build/Logging/FancyLogger/ANSIBuilder.cs]: I'm happy to wait on this!

Comment (edvilme) [reply src/Build/Logging/FancyLogger/ANSIBuilder.cs]: It is now up :)

**Review by rainersigwald — APPROVED** [review]

**Review by Forgind — APPROVED** [review]

**Comment by rainersigwald** [comment]
@edvilme can you please merge `main` after we land the 17.5 merge into main (I think doing those together will minimize merge conflicts).

**Comment by rainersigwald** [comment]
I merged #8314 into this speculatively and it's looking good but blocked on that, so drafted for now.

**Comment by rainersigwald** [comment]
Undrafted since #8314 went in. Still LGTM to me, so marking merge-when.

**Pull request merged**
